cmd/image-builder-maintenance/db.go

```diff
@@ -0,0 +1,213 @@
+		return 0, err
+	}
+	return count, nil
+}
+
+func (d *maintenanceDB) VacuumAnalyze() error {
+	_, err := d.Conn.Exec(context.Background(), sqlVacuumAnalyze)
+	if err != nil {
+		return fmt.Errorf("Error running VACUUM ANALYZE: %v", err)
+	}
+	return nil
+}
+
+func (d *maintenanceDB) LogVacuumStats() (int64, error) {
+	rows, err := d.Conn.Query(context.Background(), sqlVacuumStats)
+	if err != nil {
+		return int64(0), fmt.Errorf("Error querying vacuum stats: %v", err)
+	}
+	defer rows.Close()
+
+	deleted := int64(0)
+
+	for rows.Next() {
+		var relName, relSize string
+		var ins, upd, del, live, dead, vc, avc, ac, aac int64
+		var lvc, lavc, lan, laan *time.Time
+
+		err = rows.Scan(&relName, &relSize, &ins, &upd, &del, &live, &dead,
+			&vc, &avc, &ac, &aac,
+			&lvc, &lavc, &lan, &laan)
+		if err != nil {
+			return int64(0), err
+		}
+
+		logrus.WithFields(logrus.Fields{
+			"table_name":        relName,
+			"table_size":        relSize,
+			"tuples_inserted":   ins,
+			"tuples_updated":    upd,
+			"tuples_deleted":    del,
+			"tuples_live":       live,
+			"tuples_dead":       dead,
+			"vacuum_count":      vc,
+			"autovacuum_count":  avc,
+			"last_vacuum":       lvc,
+			"last_autovacuum":   lavc,
+			"analyze_count":     ac,
+			"autoanalyze_count": aac,
+			"last_analyze":      lan,
+			"last_autoanalyze":  laan,
+		}).Info("Vacuum and analyze stats for table")
+	}
+	if rows.Err() != nil {
+		return int64(0), rows.Err()
+	}
+	return deleted, nil
+
+}
+
+func DBCleanup(dbURL string, dryRun bool, dbEmailRetentionMonths int) error {
+	db, err := newDB(dbURL)
+	if err != nil {
+		return err
+	}
+
+	_, err = db.LogVacuumStats()
+	if err != nil {
+		logrus.Errorf("Error running vacuum stats: %v", err)
+	}
+
+	var rowsClones int64
+	var rows int64
+
+	emailRetentionDate := time.Now().AddDate(0, dbEmailRetentionMonths*-1, 0)
+
+	for {
+		if dryRun {
+			rowsClones, err = db.ExpiredClonesCount(emailRetentionDate)
+			if err != nil {
+				logrus.Warningf("Error querying expired clones: %v", err)
+			}
+
+			rows, err = db.ExpiredComposesCount(emailRetentionDate)
+			if err != nil {
+				logrus.Warningf("Error querying expired composes: %v", err)
+			}
+			logrus.Infof("Dryrun, expired composes count: %d (affecting %d clones)", rows, rowsClones)
+			break
+		}
+
+		rows, err = db.DeleteClones(emailRetentionDate)
+		if err != nil {
+			logrus.Errorf("Error deleting clones: %v, %d rows affected", rows, err)
+			return err
+		}
+
+		rows, err = db.DeleteComposes(emailRetentionDate)
+		if err != nil {
+			logrus.Errorf("Error deleting composes: %v, %d rows affected", rows, err)
+			return err
+		}
+
+		err = db.VacuumAnalyze()
+		if err != nil {
+			logrus.Errorf("Error running vacuum analyze: %v", err)
+			return err
+		}
+
+		if rows == 0 {
+			break
+		}
+
+		logrus.Infof("Deleted results for %d", rows)
+	}
+
+	_, err = db.LogVacuumStats()
+	if err != nil {
+		logrus.Errorf("Error running vacuum stats: %v", err)
+	}
+
+	return nil
+}
```

Comment:
IB projects really need to start passing `context.Context` around, it is quite powerful thing for logging. For example, if we choose to start logging into kibana, logs could be easily correlated by some kind of "cleanup_id".

In fact, I think it might be a good idea to actually configure logrus with cloudwatch because then we are getting one month history of cleanup logs for free, now once pod is gone I think logs are gone too, or at least pretty early.

Antoher benefit of passing contexts properly into the SQL driver instead of `Background()` is operational - if this process get stuck for any reason (happened in edge when we tried to delete 40mio records) you can create a deadline easily in the `main()` function and the SQL driver/server will respect that and terminate properly if the cleanup is simply too slow so they wouuld not stack up. If OpenShift kills a pod the SQL transaction will continue consuming resources on RDS, I witnessed this.

Reply:
please check